File: Ruby/correct_string.rb
def correct_char(char)
  case char
  when '5'
    return 'S'
  when '0'
    return 'O'
  when '1'
    return 'I'
  else
    return 'a'
  end
end

def misinterpreted?(char)
  return false unless ['5', '0', '1'].include?(char)
  true
end

def correct(string)
  result = ''
  string.chars.each do |char|
    if misinterpreted?(char)
      result += correct_char(char)
    else
      result += char
    end
  end
  result
end

puts correct('L0ND0N')

File: JavaScript/correctString.js
function correct(string) {
  let result = '';
  const misinterpretedCharactersMap = {
    5: 'S',
    0: 'O',
    1: 'I',
  };
  const misinterpretedCharacters = Object.keys(misinterpretedCharactersMap);
  string.split('').forEach((char) => {
    if (misinterpretedCharacters.includes(char)) {
      result += misinterpretedCharactersMap[char];
    } else {
      result += char;
    }
  });
  return result;
}

console.log(correct('L0ND0N'));

// Test.assertEquals(correct("L0ND0N"),"LONDON");
// Test.assertEquals(correct("DUBL1N"),"DUBLIN");
// Test.assertEquals(correct("51NGAP0RE"),"SINGAPORE");
// Test.assertEquals(correct("BUDAPE5T"),"BUDAPEST");
// Test.assertEquals(correct("PAR15"),"PARIS");

// S is misinterpreted as 5
// O is misinterpreted as 0
// I is misinterpreted as 1